feat: cache pre-instantiated Wasmtime modules

rvolosatovs opened the pull request:
This PR enables one of the main Wasmtime optimization strategies - pre-instantiation. Effectively, we now resolve all imports and lookup the memory index ahead-of-time and cache the result of this resolution.

To allow for that we no longer replace implicit `memory` blocks by imports when using Wasmtime - this ensures that Wasmtime manages memory "behind-the-scenes" and can do so in an optimized way utilizing e.g. it's own memory pool.
The `memory` import's type is preserved, so there should be no functional difference.

See https://docs.wasmtime.dev/examples-fast-instantiation.html#use-instancepre for more documentation on `InstancePre`

## Benchmarks

### ea33aab (main, wasmer)

```
Applied the chunk 161 times. - average stats: (application time: 115.2ms, applications per second: 8.68, gas_burned: 747.2 TGas)
```

### https://github.com/cosmonic-labs/nearcore/commit/55e03acc137466e50d22d6a57a9e08dc912ecdb8 (before)

```
Applied the chunk 174 times. - average stats: (application time: 113.0ms, applications per second: 8.85, gas_burned: 747.2 TGas)
```

### https://github.com/cosmonic-labs/nearcore/commit/55e12730ca6e02cd45e3ef3f18c74ad116ee28b6 (this PR)
```
Applied the chunk 173 times. - average stats: (application time: 102.6ms, applications per second: 9.75, gas_burned: 747.2 TGas)
```

rvolosatovs commented on runtime/near-vm-runner/src/tests/compile_errors.rs:
```diff
@@ -285,10 +285,12 @@ pub fn test_stabilized_host_function() {
 )"#,
         )
+        // WasmtimeVM uses less gas than NearVM
+        .opaque_outcome()
```
Most likely this difference is caused by a smaller module - if the module does not use memory, we do not "inject" it using Wasmtime

nagisa commented on runtime/near-vm-runner/src/wasmtime_runner.rs:
```diff
@@ -449,39 +463,31 @@ impl crate::PreparedContract for VMResult<PreparedContract> {
         let logic = store.data_mut().logic.take().expect("logic missing");
-        lazy_drop(Box::new((linker, module)));
+        drop(store);
+        lazy_drop(Box::new(instance));
```
Can you add this for now? Was a workaround necessary for wasmer but not sure if wasmtime suffers from the same issue(s) of expensive drop:

```suggestion
        // TODO: verify that lazy dropping actually improves the throughput.
        lazy_drop(Box::new(instance));
```

rvolosatovs replied:
FWIW, https://github.com/near/nearcore/pull/14060 will affect this in significant ways:
- there should not be an expensive drop, since all/most resources should be pooled internally in Wasmtime
- delaying the drop would mean delaying the release of the execution slot (refs the whole semaphore discussion https://github.com/near/nearcore/pull/14060#issuecomment-3183731514)

nagisa commented on runtime/near-vm-runner/src/wasmtime_runner.rs:
```diff
@@ -233,52 +239,63 @@ impl WasmtimeVM {
+                // we load what we think we load.
+                let module = unsafe { Module::deserialize(&self.engine, &module) }
+                    .map_err(|err| VMRunnerError::LoadingError(err.to_string()))?;
+                let memory = module.get_export_index("memory");
```
From what I can tell, since `prepare_v*` does not export the memory by default (did I overlook something?) this will be returning a `None` in most cases, won't it?

Wouldn't it be better to either always make the memory exported or always rely on the code that works with non-exported memories?

rvolosatovs replied:
`prepare` actually passes the exports through https://github.com/near/nearcore/blob/5865668aa6d8cea15eafd5e4886eaaf70dafaf50/runtime/near-vm-runner/src/prepare/prepare_v3.rs#L103-L108.
Conventionally, modules using memory export it as `memory` symbol, which is why, effectively, this should only ever return `None` for modules, which do not use memory.

I was trying to find the `"memory"` symbol name standardized somewhere, but it did find quite a few examples in

the spec:
- https://github.com/WebAssembly/spec/blob/1041527d508218acc40f0278d4abc3be9ba5e3bd/proposals/bulk-memory-operations/Overview.md?plain=1#L91
- https://github.com/WebAssembly/spec/blob/1041527d508218acc40f0278d4abc3be9ba5e3bd/interpreter/README.md?plain=1#L452
- https://github.com/WebAssembly/spec/blob/1041527d508218acc40f0278d4abc3be9ba5e3bd/test/core/memory.wast#L247
- https://github.com/WebAssembly/spec/blob/1041527d508218acc40f0278d4abc3be9ba5e3bd/test/core/memory_grow.wast#L313

MDN:
- https://developer.mozilla.org/en-US/docs/WebAssembly/Reference/Memory/Store
- https://developer.mozilla.org/en-US/docs/WebAssembly/Reference/Memory/Load

If you want, we could:
- skip all *memory* exports
- (for Wasmtime) if `memory` section found, export it as `memory`

This would ensure that if a module is using memory, it always exports it as `memory`

rvolosatovs replied:
also updated the test in https://github.com/near/nearcore/pull/14057/commits/3ad6b0804f6b2e6c7b645fd16fdb790f88f8a17d to ensure this indeed works

nagisa replied:
> I was trying to find the "memory" symbol name standardized somewhere, but it did failed to do so. I did, however, find quite a few examples in the spec:

Well `(memory (export "memory") 1)` is explicitly exporting the memory.

>  Conventionally, modules using memory export it as memory symbol, which is why, effectively, this should only ever return None for modules, which do not use memory.

Yeah, but that's just a convention. I actually wasn't aware of it! It seems to me that nothing really prevents somebody from writing a module by hand that does not have such an export, and we aren't going out of our way to add such an export either (if it is missing.)

Other risk factor I see that there might already be contracts out there that export something else (e.g. functions) named `"memory"` and we can't really go and break those. It technically is possible to go and look at all contracts to verify whether that is(n't) the case, but it seems to me like it would be better for us to avoid that risk altogether.

Unfortunately the only place I can see that enables grabbing non-exported memories at all is in `wasmtime::runtime::vm::instance::Instance` – which is an implementation detail within wasmtime :( Even though the memory itself that's stored inside the `Store` does not differ much.

I guess a question is how other conventional wasmtime embedders go about obtaining the memory in their host functions if that memory wasn't imported and/or exported? Or is that simply not something that's possible?

rvolosatovs replied:
Conventionally, embedders get access to the memory by getting the `memory` export, as seen in the example e.g. here: https://docs.rs/wasmtime/latest/wasmtime/struct.Caller.html#method.get_module_export or here: https://docs.wasmtime.dev/examples-rust-memory.html

nagisa commented:
Lets make sure to add some unit tests along the lines of:

1. Module does not export any memory at all (and then calls some host function that uses memory);
2. Module exports a memory but under a different name than `"memory"`;
3. Module exports something else with `"memory"` as a name…

rvolosatovs commented:
What do you think about `___nearcore_memory` approch, @nagisa (https://github.com/near/nearcore/pull/14057/commits/c8bb7ecff1f76b7094ccebb9b997b2f024ab10f8)?

I'll follow-up with the test case, but that'll likely have to wait for tomorrow

I do want to point out that the `test_stabilized_host_function` test case is already one that uses a function, which utilizes memory, but neither defines it, nor exports it

rvolosatovs commented:
@nagisa I believe I've addressed all the feedback. As I pointed out earlier,

> Input module does not export any memory at all (and then calls some host function that uses memory);

case is already covered by `test_stabilized_host_function`, so I've addressed the other two test case scenarios
